### ISSUE TITLE
sddm-greeter process dies if Qt is build in debug mode

### ISSUE THREAD

**Issue by NigridsVa** [sddm/sddm#506]
I run SDDM 0.12 on Arch Linux (the only change to vanilla being the patch contained in commit c637727) and recompiled Qt 5.5.0 with debug support enabled.

When SDDM is started from systemd, the subprocess sddm-greeter dies because a Qt assertion is in triggered. Here is the relevant excerpt from journalctl:

```
Okt 03 11:39:22 systemd[1]: Started Simple Desktop Display Manager.
Okt 03 11:39:22 sddm[12091]: Initializing...
Okt 03 11:39:22 sddm[12091]: Starting...
Okt 03 11:39:22 sddm[12091]: Adding new display on vt 1 ...
Okt 03 11:39:22 sddm[12091]: Display server starting...
Okt 03 11:39:22 sddm[12091]: Running: /usr/bin/X -nolisten tcp -auth /var/run/sddm/{99f8f7a5-8568-4cdf-a801-ac701e60cbf3} -background none -noreset -displayfd 17 vt1
Okt 03 11:39:22 polkitd[429]: Unregistered Authentication Agent for unix-process:12085:418554 (system bus name :1.119, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale de_DE.UTF-8) (disconnected from bus)
Okt 03 11:39:22 sddm[12091]: Running display setup script  "/usr/share/sddm/scripts/Xsetup"
Okt 03 11:39:22 sddm[12091]: Display server started.
Okt 03 11:39:22 sddm[12091]: Socket server starting...
Okt 03 11:39:22 sddm[12091]: Socket server started.
Okt 03 11:39:22 sddm[12091]: Greeter starting...
Okt 03 11:39:22 sddm[12091]: Adding cookie to "/var/run/sddm/{99f8f7a5-8568-4cdf-a801-ac701e60cbf3}"
Okt 03 11:39:22 sddm-helper[12097]: [PAM] Starting...
Okt 03 11:39:22 sddm-helper[12097]: [PAM] Authenticating...
Okt 03 11:39:22 sddm-helper[12097]: [PAM] returning.
Okt 03 11:39:22 sddm-helper[12097]: pam_unix(sddm-greeter:session): session opened for user sddm by (uid=0)
Okt 03 11:39:22 systemd[1]: Created slice user-994.slice.
Okt 03 11:39:22 systemd[1]: Starting User Manager for UID 994...
Okt 03 11:39:22 systemd-logind[375]: New session c6 of user sddm.
Okt 03 11:39:22 systemd[1]: Started Session c6 of user sddm.
Okt 03 11:39:22 systemd[12099]: pam_unix(systemd-user:session): session opened for user sddm by (uid=0)
Okt 03 11:39:22 systemd[12099]: Listening on D-Bus User Message Bus Socket.
Okt 03 11:39:22 systemd[12099]: Reached target Paths.
Okt 03 11:39:22 systemd[12099]: Listening on Sound System.
Okt 03 11:39:22 systemd[12099]: Reached target Timers.
Okt 03 11:39:22 systemd[12099]: Reached target Sockets.
Okt 03 11:39:22 systemd[12099]: Reached target Basic System.
Okt 03 11:39:22 systemd[12099]: Reached target Default.
Okt 03 11:39:22 systemd[12099]: Startup finished in 11ms.
Okt 03 11:39:22 systemd[1]: Started User Manager for UID 994.
Okt 03 11:39:22 sddm[12091]: Greeter session started successfully
Okt 03 11:39:22 sddm-greeter[12103]: Reading from "/usr/share/xsessions/plasma.desktop"
Okt 03 11:39:22 sddm-greeter[12103]: Reading from "/usr/share/xsessions/failsafe.desktop"
Okt 03 11:39:22 sddm-greeter[12103]: Connected to the daemon.
Okt 03 11:39:22 sddm[12091]: Message received from greeter: Connect
Okt 03 11:39:22 sddm-greeter[12103]: QObject: Cannot create children for a parent that is in a different thread.
                                             (Parent is SDDM::GreeterApp(0x7fffd19b28a0), parent's thread is QThread(0xaf5830), current thread is QThread(0xb81040)
Okt 03 11:39:22 sddm-greeter[12103]: ASSERT failure in QCoreApplication::sendEvent: "Cannot send events to objects owned by a different thread. Current thread b81040. Receiver '' (of type 'SDDM::GreeterApp') was created in thread af5830", file kernel/qcoreapplication.cpp, line 553
Okt 03 11:39:22 sddm-helper[12097]: [PAM] Ended.
Okt 03 11:39:22 sddm[12091]: Auth: sddm-helper exited with 6
Okt 03 11:39:22 sddm[12091]: Greeter stopped.
```

The backtrace contained in the produced coredump looks like the following:

```
#0  0x00007fc2c126b5f8 in raise () from /usr/lib/libc.so.6
#1  0x00007fc2c126ca7a in abort () from /usr/lib/libc.so.6
#2  0x00007fc2c1c1c9f6 in qt_message_fatal (context=..., message=<synthetic pointer>) at global/qlogging.cpp:1575
#3  QMessageLogger::fatal (this=this@entry=0x7fc2ac2092d0, msg=msg@entry=0x7fc2c1eb5010 "ASSERT failure in %s: \"%s\", file %s, line %d") at global/qlogging.cpp:777
#4  0x00007fc2c1c171a6 in qt_assert_x (where=where@entry=0x7fc2c1f62a32 "QCoreApplication::sendEvent", what=<optimized out>, file=file@entry=0x7fc2c1f629e4 "kernel/qcoreapplication.cpp", line=line@entry=553) at global/qglobal.cpp:2978
#5  0x00007fc2c1dffb5f in QCoreApplicationPrivate::checkReceiverThread (this=this@entry=0xaf72c0, receiver=receiver@entry=0x7fffd19b28a0) at kernel/qcoreapplication.cpp:545
#6  0x00007fc2c1e003c2 in QCoreApplication::notify (this=<optimized out>, receiver=0x7fffd19b28a0, event=0x7fc2ac2094e0) at kernel/qcoreapplication.cpp:1035
#7  0x00007fc2c1e0002d in QCoreApplication::notifyInternal (this=0x7fffd19b28a0, receiver=0x7fffd19b28a0, event=event@entry=0x7fc2ac2094e0) at kernel/qcoreapplication.cpp:965
#8  0x00007fc2c1e071c8 in QCoreApplication::sendEvent (event=0x7fc2ac2094e0, receiver=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:224
#9  QCoreApplication::installTranslator (translationFile=<optimized out>) at kernel/qcoreapplication.cpp:1843
#10 0x00007fc29edcac7d in ?? () from /usr/lib/libKF5SonnetUi.so.5
#11 0x00007fc29edcae3d in ?? () from /usr/lib/libKF5SonnetUi.so.5
#12 0x00007fc2c1e04282 in qAddPreRoutine (p=0x7fc29edcac90) at kernel/qcoreapplication.cpp:245
#13 0x00007fc2c3e5027a in call_init.part () from /lib64/ld-linux-x86-64.so.2
#14 0x00007fc2c3e5038b in _dl_init () from /lib64/ld-linux-x86-64.so.2
#15 0x00007fc2c3e54830 in dl_open_worker () from /lib64/ld-linux-x86-64.so.2
#16 0x00007fc2c3e50124 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#17 0x00007fc2c3e53fd1 in _dl_open () from /lib64/ld-linux-x86-64.so.2
#18 0x00007fc2c092cfc9 in ?? () from /usr/lib/libdl.so.2
#19 0x00007fc2c3e50124 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#20 0x00007fc2c092d5e1 in ?? () from /usr/lib/libdl.so.2
#21 0x00007fc2c092d061 in dlopen () from /usr/lib/libdl.so.2
#22 0x00007fc2c1dfc774 in QLibraryPrivate::load_sys (this=0x7fc2a4060180) at plugin/qlibrary_unix.cpp:227
#23 0x00007fc2c1df64b5 in QLibraryPrivate::load (this=this@entry=0x7fc2a4060180) at plugin/qlibrary.cpp:534
#24 0x00007fc2c1df673b in QLibraryPrivate::loadPlugin (this=0x7fc2a4060180) at plugin/qlibrary.cpp:582
#25 0x00007fc2c1ded103 in QPluginLoader::load (this=this@entry=0x7fc2a404f4d0) at plugin/qpluginloader.cpp:232
#26 0x00007fc2c28dc281 in QQmlImportDatabase::importDynamicPlugin (this=this@entry=0xb80f48, filePath=..., uri=..., typeNamespace=..., vmaj=vmaj@entry=2, errors=errors@entry=0x7fc2ac20a660) at qml/qqmlimport.cpp:1973
#27 0x00007fc2c28dd9cc in QQmlImportsPrivate::importExtension (this=this@entry=0xcded00, qmldirFilePath=..., uri=..., vmaj=vmaj@entry=2, vmin=vmin@entry=0, database=database@entry=0xb80f48, qmldir=0x7fc2a405fca0, errors=0x7fc2ac20a660)
    at qml/qqmlimport.cpp:927
#28 0x00007fc2c28de6c5 in QQmlImportsPrivate::addLibraryImport (this=0xcded00, uri=..., prefix=..., vmaj=vmaj@entry=2, vmin=vmin@entry=0, qmldirIdentifier=..., qmldirUrl=..., incomplete=false, database=0xb80f48, errors=0x7fc2ac20a660)
    at qml/qqmlimport.cpp:1266
#29 0x00007fc2c28de9b3 in QQmlImports::addLibraryImport (this=this@entry=0xce7f90, importDb=importDb@entry=0xb80f48, uri=..., prefix=..., vmaj=2, vmin=0, qmldirIdentifier=..., qmldirUrl=..., incomplete=false, errors=0x7fc2ac20a660)
    at qml/qqmlimport.cpp:1497
#30 0x00007fc2c28a79fa in QQmlTypeLoader::Blob::addImport (this=this@entry=0xce7f30, import=import@entry=0x7fc2a4019000, errors=errors@entry=0x7fc2ac20a660) at qml/qqmltypeloader.cpp:1330
#31 0x00007fc2c28a7edd in QQmlTypeData::continueLoadFromIR (this=this@entry=0xce7f30) at qml/qqmltypeloader.cpp:2172
#32 0x00007fc2c28a88fd in QQmlTypeData::dataReceived (this=0xce7f30, data=...) at qml/qqmltypeloader.cpp:2129
#33 0x00007fc2c28a23e6 in QQmlTypeLoader::setData (this=this@entry=0xb80eb8, blob=blob@entry=0xce7f30, d=...) at qml/qqmltypeloader.cpp:1199
#34 0x00007fc2c28a2e4f in QQmlTypeLoader::setData (this=this@entry=0xb80eb8, blob=0xce7f30, file=file@entry=0x7fc2ac20a910) at qml/qqmltypeloader.cpp:1191
#35 0x00007fc2c28a31f3 in QQmlTypeLoader::loadThread (this=0xb80eb8, blob=blob@entry=0xce7f30) at qml/qqmltypeloader.cpp:1069
#36 0x00007fc2c28a3624 in QQmlTypeLoaderThread::loadThread (this=<optimized out>, b=0xce7f30) at qml/qqmltypeloader.cpp:813
#37 0x00007fc2c290cc53 in QQmlThreadPrivate::threadEvent (this=0xb81040) at qml/ftw/qqmlthread.cpp:190
#38 0x00007fc2c290d4a1 in QQmlThreadPrivate::event (this=0xb81040, e=0xcded60) at qml/ftw/qqmlthread.cpp:128
#39 0x00007fc2c1e00406 in QCoreApplication::notify (this=<optimized out>, receiver=0xb81040, event=0xcded60) at kernel/qcoreapplication.cpp:1038
#40 0x00007fc2c1e0002d in QCoreApplication::notifyInternal (this=0x7fffd19b28a0, receiver=0xb81040, event=event@entry=0xcded60) at kernel/qcoreapplication.cpp:965
#41 0x00007fc2c1e0393c in QCoreApplication::sendEvent (event=0xcded60, receiver=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:224
#42 QCoreApplicationPrivate::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0, data=0xb811f0) at kernel/qcoreapplication.cpp:1593
#43 0x00007fc2c1e03dbc in QCoreApplication::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0) at kernel/qcoreapplication.cpp:1451
#44 0x00007fc2c1e5974a in postEventSourceDispatch (s=0x7fc2a40012d0) at kernel/qeventdispatcher_glib.cpp:271
#45 0x00007fc2bd842dc7 in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#46 0x00007fc2bd843020 in ?? () from /usr/lib/libglib-2.0.so.0
#47 0x00007fc2bd8430cc in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
#48 0x00007fc2c1e597eb in QEventDispatcherGlib::processEvents (this=0x7fc2a40008c0, flags=...) at kernel/qeventdispatcher_glib.cpp:420
#49 0x00007fc2c1dfe813 in QEventLoop::exec (this=this@entry=0x7fc2ac20ad90, flags=..., flags@entry=...) at kernel/qeventloop.cpp:204
#50 0x00007fc2c1c2a4e4 in QThread::exec (this=this@entry=0xb81040) at thread/qthread.cpp:503
#51 0x00007fc2c290c829 in QQmlThreadPrivate::run (this=0xb81040) at qml/ftw/qqmlthread.cpp:141
#52 0x00007fc2c1c301fc in QThreadPrivate::start (arg=0xb81040) at thread/qthread_unix.cpp:331
#53 0x00007fc2bf7c64a4 in start_thread () from /usr/lib/libpthread.so.0
#54 0x00007fc2c132113d in clone () from /usr/lib/libc.so.6
```

**Comment by davispuh** [comment]
Did you recompile sddm after Qt compilation? I noticed that when you compile Qt must also recompile most of Qt apps...

**Comment by NigridsVa** [comment]
At first I had not done this, but I thought, perhaps a debug build of sddm will give better result. No change. I just recompiled it again, as Arch introduced a patch contained in 4cfed6b, but still the same error message.

Today Qt 5.5.1 was released. Perhaps I should give it a try and see if the error still persists?

**Comment by davispuh** [comment]
I just found out why it happens, you can't mix different Qt builds and it's not enough to rebuild just Qt :(

In `/etc/sddm.conf` set 
```
[Theme]
Current=maui
```

Then for sddm need to recompile only

```
qt5-base
qt5-declarative
qt5-xmlpatterns
```

But from your log, I see you use Plasma
```
10 0x00007fc29edcac7d in ?? () from /usr/lib/libKF5SonnetUi.so.5
```

so you'll need to recompile all these

```
qt5-base
qt5-declarative
qt5-xmlpatterns
qt5-svg
sonnet
kcoreaddons
kauth
kconfig
kwidgetsaddons
kwidgetsaddons
kcompletion
kjobwidgets
kjobwidgets
kcodects
kcodecs
kitemviews
kwindowsystem
kwindowsystem
knotifications
kglobalaccel
```

can see Plasma libs are dynamically loaded
```
#11 0x00007fc29edcae3d in ?? () from /usr/lib/libKF5SonnetUi.so.5
[...]
#21 0x00007fc2c092d061 in dlopen () from /usr/lib/libdl.so.2
```
(after recompiling sonnet you'll see different lib)

You can recompile them with ABS or they all are in AUR as git versions which makes it quite easy to recompile.
But note that this is only to get sddm working, Plasma desktop might not work as it might need more libraries to be recompiled. Probably easiest way is just install `plasma-git-meta` from AUR.

**Comment by NigridsVa** [comment]
The interesting bit is, that KDE still worked without recompiling anything. I just disabled sddm and used the good old startx script for starting the desktop environment. But you think, the actual error might be caused from the libKF5SonnetUI library?

**Comment by davispuh** [comment]
yeah, I tested it myself. I recompiled only `qt5-base` and then I got

``
Cannot mix incompatible Qt library (version 0x50500) with this library (version 0x50700)
``

then I recompiled `qt5-declarative, qt5-xmlpatterns` and changed theme to `maui` which made sddm to work, but if I set `breeze` theme I got this exactly same crash
```
sddm-greeter: ASSERT failure in QCoreApplication::sendEvent: "Cannot send events to objects owned by a different thread. Current thread 118aa50. Receiver '' (of type 'SDDM::GreeterApp') was created in thread 111fc20", file /mnt/Qt/qtbase/src/corelib/kernel/qcoreapplication.cpp, line 556
```

in backtrack I saw same `libKF5SonnetUi` like you and then I recompiled Sonnet which madr next backtrace with same error message but instead of Sonnet it had `libKF5CoreAddons` and so on. After recompiling those all libs sddm worked.
